### ISSUE TITLE
Docker container

### ISSUE THREAD

**Issue by espenhgn** [LFPy/LFPy#146]
Docker is being used here, there and everywhere it seems. Adding a Dockerfile to build a container containing LFPy should be fairly straightforward.

**Comment by torbjone** [comment]
I'm new to this, but I recently talked to Sebastian Spreizer who has developed the NEST Desktop, and for that case, if you have docker on you computer you can just write "docker run -i -p 5000:5000 -p 8000:8000 -t babsey/nest-desktop" and this single line is enough to download and install everything. Maybe we could make this work for LFPy as well? We would probably have to register at DockerHub: 
(https://hub.docker.com/r/babsey/nest-desktop)

**Comment by espenhgn** [comment]
Hi, that may also work if we want to maintain another remote account. My Dockerfile just installs a few packages in a preexisting docker image (https://hub.docker.com/r/continuumio/miniconda3). Btw., I now found you can build from URLs, and updated the README in #147 accordingly:
```
    docker build -t lfpy https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile
    docker run -it -p 8888:8888 lfpy
```

**Comment by torbjone** [comment]
When I try to run the first line, I get:
tone@tone-latitude-e5450:~$ docker build -t lfpy https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile
unable to prepare context: unable to download remote context https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile: failed to GET https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile with status 404 Not Found: 404: Not Found
tone@tone-latitude-e5450:~$ docker build -t lfpy https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile
unable to prepare context: unable to download remote context https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile: failed to GET https://raw.githubusercontent.com/LFPy/LFPy/docker/Dockerfile with status 404 Not Found: 404: Not Found

**Comment by espenhgn** [comment]
Hi, it's not merged yet, so you'd have to point to the working branch on my user ;) 
Try: 
```
docker build -t lfpy https://raw.githubusercontent.com/espenhgn/LFPy/docker/Dockerfile
docker run -it -p 8888:8888 lfpy
```

**Comment by torbjone** [comment]
*facepalm*

I get an error, but I will try to look at it and see if I can resolve it:

tone@tone-latitude-e5450:~$ docker build -t lfpy https://raw.githubusercontent.com/espenhgn/LFPy/docker/Dockerfile
Downloading build context from remote url: https://raw.githubusercontent.com/espenhgn/LFPy/docker/Dockerfile     356B
Sending build context to Docker daemon  2.048kB
Step 1/8 : FROM continuumio/miniconda3
 ---> f700f7f570c7
Step 2/8 : RUN conda config --add channels conda-forge
 ---> Using cache
 ---> ea549fd7e36c
Step 3/8 : RUN conda config --set channel_priority strict
 ---> Running in 90b777c020c5
An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : linux-64
          conda version : 4.3.27
       conda is private : False
      conda-env version : 4.3.27
    conda-build version : not installed
         python version : 3.6.2.final.0
       requests version : 2.18.4
       root environment : /opt/conda  (writable)
    default environment : /opt/conda
       envs directories : /opt/conda/envs
                          /root/.conda/envs
          package cache : /opt/conda/pkgs
                          /root/.conda/pkgs
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.continuum.io/pkgs/main/linux-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/linux-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : /root/.condarc
             netrc file : None
           offline mode : False
             user-agent : conda/4.3.27 requests/2.18.4 CPython/3.6.2 Linux/4.15.0-62-generic debian/8.9 glibc/2.19    
                UID:GID : 0:0

`$ /opt/conda/bin/conda config --set channel_priority strict`




    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/conda/_vendor/auxlib/type_coercion.py", line 164, in boolify
        return bool(complex(val))
    ValueError: complex() arg is a malformed string
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/conda/_vendor/auxlib/type_coercion.py", line 243, in typify
        return boolify(value) if type_hint == bool else type_hint(value)
      File "/opt/conda/lib/python3.6/site-packages/conda/_vendor/auxlib/type_coercion.py", line 168, in boolify
        raise TypeCoercionError(value, "The value %r cannot be boolified." % value)
    conda._vendor.auxlib.type_coercion.TypeCoercionError: The value 'strict' cannot be boolified.
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/opt/conda/lib/python3.6/site-packages/conda/exceptions.py", line 640, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/opt/conda/lib/python3.6/site-packages/conda/cli/main.py", line 140, in _main
        exit_code = args.func(args, p)
      File "/opt/conda/lib/python3.6/site-packages/conda/cli/main_config.py", line 210, in execute
        execute_config(args, parser)
      File "/opt/conda/lib/python3.6/site-packages/conda/cli/main_config.py", line 402, in execute_config
        value = context.typify_parameter(key, item)
      File "/opt/conda/lib/python3.6/site-packages/conda/common/configuration.py", line 904, in typify_parameter
        return typify_data_structure(value, parameter._element_type)
      File "/opt/conda/lib/python3.6/site-packages/conda/_vendor/auxlib/type_coercion.py", line 259, in typify_data_structure
        return typify(value, type_hint)
      File "/opt/conda/lib/python3.6/site-packages/conda/_vendor/auxlib/type_coercion.py", line 246, in typify
        raise TypeCoercionError(value, text_type(e))
    conda._vendor.auxlib.type_coercion.TypeCoercionError: The value 'strict' cannot be boolified.

The command '/bin/sh -c conda config --set channel_priority strict' returned a non-zero code: 1

**Comment by espenhgn** [comment]
That's weird. If I understood things correctly, the container shouldn't depend on the host OS at all. For me it works like:
```
$ docker build -t lfpy https://raw.githubusercontent.com/espenhgn/LFPy/docker/Dockerfile
Downloading build context from remote url: https://raw.githubusercontent.com/espenhgn/LFPy/docker/Dockerfile     356B
Sending build context to Docker daemon  2.048kB
Step 1/8 : FROM continuumio/miniconda3
 ---> 4a51de2367be
Step 2/8 : RUN conda config --add channels conda-forge
 ---> Using cache
 ---> 93c70f878924
Step 3/8 : RUN conda config --set channel_priority strict
 ---> Using cache
 ---> 6888bca5f553
Step 4/8 : RUN conda create -n lfpy python=3.7 lfpy ipython jupyter git matplotlib make gxx_linux-64
 ---> Running in 5c601f2da0a4
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... done

....
```

**Comment by espenhgn** [comment]
If I can suggest something, try running the container "continuumio/miniconda3" directly and repeat the steps in the Dockerfile I set up and figure out why it fails with `conda config --set channel_priority strict`

**Comment by espenhgn** [comment]
Hi @torbjone . Did you figure out why the build recipe failed on your end? I'm a bit hesitant to merge  PR #147 ...

**Comment by torbjone** [comment]
No, not really. The NEST Desktop docker works perfectly on both my laptop and desktop, but I can not make the LFPy docker work on either of them.

**Comment by torbjone** [comment]
On my desktop, I am able to build and run, but I only enter a terminal, there is no link I can open in a browser or anything.

tone@think:~/dockertest$ docker run -it -p 8888:8888 lfpy
(lfpy) root@1529e9b15bda:/# 

If I try to open a jupter notebook, I get:
(lfpy) root@1529e9b15bda:/# jupyter-notebook
[I 19:34:51.308 NotebookApp] Writing notebook server cookie secret to /root/.local/share/jupyter/runtime/notebook_cookie_secret
Traceback (most recent call last):
  File "/opt/conda/envs/lfpy/bin/jupyter-notebook", line 11, in <module>
    sys.exit(main())
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/jupyter_core/application.py", line 267, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/traitlets/config/application.py", line 663, in launch_instance
    app.initialize(argv)
  File "</opt/conda/envs/lfpy/lib/python3.7/site-packages/decorator.py:decorator-gen-7>", line 2, in initialize
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/notebook/notebookapp.py", line 1679, in initialize
    self.init_webapp()
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/notebook/notebookapp.py", line 1442, in init_webapp
    self.http_server.listen(port, self.ip)
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/tornado/tcpserver.py", line 151, in listen
    sockets = bind_sockets(port, address=address)
  File "/opt/conda/envs/lfpy/lib/python3.7/site-packages/tornado/netutil.py", line 174, in bind_sockets
    sock.bind(sockaddr)
OSError: [Errno 99] Cannot assign requested address
(lfpy) root@1529e9b15bda:/#

**Comment by espenhgn** [comment]
I think the line `OSError: [Errno 99] Cannot assign requested address` tries to tell that the port number (8888) is not available. Can you try the same port numbers that works for the nest-desktop, like `docker run -it -p 5000:5000 -p 8000:8000 lfpy`

**Comment by torbjone** [comment]
I tried that, but got same error.
By running in the docker: 
jupyter notebook --ip=0.0.0.0 --port=8080 --allow-root
it looks like a notebook starts, and I get some links, but I can't open any of them.

**Comment by torbjone** [comment]
Actually, by running: 
jupyter notebook --ip=0.0.0.0 --port=5000 --allow-root
one of the links now works.

**Comment by espenhgn** [comment]
Great. I'm guessing the available port numbers are somewhat OS dependent. I'll test if port 5000 works on my end.

**Comment by torbjone** [comment]
Same solution works on my laptop, after commenting out the line
RUN conda config --set channel_priority strict

**Comment by torbjone** [comment]
In the description it says: "Various LFPy example files can be found in the folder ``/opt/LFPy/examples/``"

but here, the opt folder only has a 'conda' folder, while the examples are found here: /LFPy/examples/

**Comment by espenhgn** [comment]
Merged PR #147, after changing the port number and removing the line `RUN conda config --set channel_priority strict` in the build recipe. Hopefully it's now working for all.

**Comment by espenhgn** [comment]
Opened PR #149

**Comment by torbjone** [comment]
Seems to work fine, go ahead.

**Comment by espenhgn** [comment]
Merged #149 . Hopefully that does it.
